[Review] 리뷰 생성 기능 구현

rlqor9688 opened the pull request:
[Review] 리뷰 생성 기능 추가

## Description
- 리뷰 생성 시 Users, Orders, Stores, OrderItems 테이블을 참조합니다.
- 인증/인가 과정은 다음과 같습니다.
1. 로그인한 유저가 고객(CONSUMER)인지 확인 > 아닐 시 INVALID_USER_ROLE
2. ReviewReqeustDto에 담긴 ordersId와 storesId가 존재하는지 체크 -> 미 존재 시 각각 ORDER/STORE_NOT_FOUND
**3. OrderStatus가 LIVE인지 체크 -> LIVE가 아닐 시 REVIEW_ORDER_INVALID  (추후 OrderStatus Enum 명 변경 가능)
4. ReviewReqeustDto에 담긴 ordersId로 조회한 userId와 authUserId가 일치하는지 검증 > 미일치 시 REVIEW_USER_MISMATCH
5. ReviewReqeustDto에 담긴 ordersId로 조회한 storeId와 ReviewReqeustDto에 담긴 storesId가 일치하는지 검증 > 미일치 시 REVIEW_STORE_MISMATCH
6. ReviewRequestDto에 담긴 orderId로 조회한 orderItems가 없을 때 >REVIEW_ORDERITEM_NOT_FOUND
7. 위의 인증/인가를 통과한 경우 새로운 리뷰를 생성하고 Entity Table에 저장합니다.
8. ReviewResponseDto로 반환합니다.
9. Controller에서 ReviewResponseDto와 응답을 받아 반환합니다.

## Changes
- ReviewController
- ReviewRequestDto
- ReviewResponseDto
- ReviewService (인터페이스)
- ReviewServiceImpl (클래스)
- ReviewRepository JpaRepository<Reviews, Long> 상속 추가
- ReviewServiceImpl Test
- messages.properties 추가
- ErrorCode 추가
- OrderItems / OrderItemsRepository는 설정했으나 Order 작업자와의 코드 충돌 우려로 커밋 미포함

## Ref
![image](https://github.com/user-attachments/assets/c67e00ff-985b-4022-a21a-d796c6f972d1)

## 리뷰 정상 생성 테스트
![image](https://github.com/user-attachments/assets/2bc51c81-0e72-4b86-94fd-62aa44e19b91)

rlqor9688 committed:
WIP: 작업 중인 변경사항 커밋

rlqor9688 committed:
fix: Resolve merge conflicts with dev

rlqor9688 committed:
feat [review] : review 생성 기능 구현

1. ReviewController 내 createReview 메서드 구현
2. ReviewService 내 createReview 메서드 구현
3. ErrorCode 추가
4. GlobalException 생성자 추가
5. GlobalExceptionHandler Auth 메서드 오타 수정

rlqor9688 committed:
fix [review] : review 생성 기능 수정

1. ReviewRepository JpaRepository 상속 추가
2. ReviewService @AllArgsConstructor Annotation 추가

rlqor9688 committed:
Merge branch 'dev' of https://github.com/rlqor9688/igo_delivery into dev

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-review

rlqor9688 committed:
Merge branch 'dev' of https://github.com/rlqor9688/igo_delivery into dev

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-review

rlqor9688 committed:
Merge branch 'dev' of https://github.com/rlqor9688/igo_delivery into dev

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-review

rlqor9688 committed:
Merge branch 'dev' of https://github.com/rlqor9688/igo_delivery into dev

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-review

rlqor9688 committed:
feat [review] : review 생성 기능 구현

1. ReviewRepository JpaRepository 상속 추가
2. ReviewService @AllArgsConstructor Annotation 추가
3. ErrorCode 추가
4. messages.properties 추가
5. Review Controller 내 createReview 메서드 생성
6. ReviewRequestDto 추가
7. ReviewResponseDto 추가
8. ReviewServiceImpl 클래스 생성 & 클래스 내 createReview 메서드 추가
9. ReviewServiceImplTest 클래스 생성 & 리뷰 정상 생성 테스트 추가
10. OrderItemsRepository, OrderRepository는 필요 기능 구현했으나 충돌 우려로 커밋하지 않음

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/dto/ReviewResponseDto.java:
```diff
@@ -0,0 +1,27 @@
+    private String content;
+    private LocalDateTime createdAt;
+
+    public static ReviewResponseDto of(Reviews reviews) {
+        return new ReviewResponseDto(
+                reviews.getUsers().getId(),
+                reviews.getRating(),
+                reviews.getContent(),
+                reviews.getCreatedAt()
+        );
+    }
```
이부분도 세영님 PR에서 말씀 드린바와 같이 관례에 따라 from으로 변경해주시면 좋을 것 같습니다.

어떤 대상을 하나로 변환할 때는 from, 여러 값을 조합해서 새로운 인스턴스를 만들 때는 of를 사용하는 것이 일반적이라고 합니다.

rlqor9688 replied:
반영 완료

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/entity/Reviews.java:
```diff
@@ -54,5 +54,17 @@ public class Reviews extends BaseEntity {
+        this.reviewStatus = ReviewStatus.DELETED;
+    }
+
+    public static Reviews of(Users user, Orders order, Stores store, ReviewRequestDto requestDto) {
+
+        return Reviews.builder()
+                .users(user)
+                .orders(order)
+                .stores(store)
+                .content(requestDto.getContent())
+                .rating(requestDto.getRating())
+                .build();
```
여기가 of 사용의 좋은 예시 인 것 같습니다

rlqor9688 replied:
오! 이해 확 됐습니다!

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+    @Transactional(readOnly = true)
+    public ReviewResponseDto createReview(AuthUser authUser, ReviewRequestDto requestDto) {
+        // 로그인한 유저가 고객(CONSUMER)인지 확인
+        if (!Objects.equals(authUser.getUserRole(), UserRole.CONSUMER)) {
+            throw new GlobalException(ErrorCode.INVALID_USER_ROLE);
+        }
+
```
이부분 Users Entity에 메서드 만들어 두었습니다. 
유저를 먼저 찾고, 찾은 유저에게서 검증 메서드를 호출하면 해당 메서드를 활용할 수 있을 것 같습니다.

다만 지금도 동작은 정상적으로 될 것 같아서 선택적으로 적용해주시면 될 것 같네요

rlqor9688 replied:
오 감사합니다! 반영 완료

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+        }
+
+        // 유저, 주문, 가게 정보 DB에서 찾기
+        Users findUser = userRepository.findById(authUser.getId())
+                .orElseThrow(() -> new AuthException(ErrorCode.USER_NOT_FOUND));
```
요건 GlobalException으로 변경 해주시면 좋을 것 같아요, 로그인, 회원가입 같은 인증시에면 AuthExceptin 사용하도록 통일하면 좋을 것 같습니다.

rlqor9688 replied:
반영 완료

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+    private final UserRepository userRepository;
+
+    @Override
+    @Transactional(readOnly = true)
```
readOnly 설정은 조회 하는 경우에 적합해서 리뷰를 생성하는 이 메서드에는 적합하지 않은 것 같아요!

rlqor9688 replied:
감사합니다! 안 그래도 찾아봤더니 DB에 변경이 생기는 로직이 하나라도 있으면 ReadOnly를 쓰면 안된다고 하더라고요. 반영 완료!

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+    @Transactional(readOnly = true)
+    public ReviewResponseDto createReview(AuthUser authUser, ReviewRequestDto requestDto) {
+        // 로그인한 유저가 고객(CONSUMER)인지 확인
+        if (!Objects.equals(authUser.getUserRole(), UserRole.CONSUMER)) {
```
이 부분은 User 엔티티에 존재하는 validateConsumer 메서드를 활용하는 것이 역할 분리가 되어 좋을 것 같아요!

rlqor9688 replied:
감사합니다! 반영 완료

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+            throw new GlobalException(ErrorCode.REVIEW_ORDERITEM_NOT_FOUND);
+        }
+
+        // 리뷰 생성
+        Reviews review = Reviews.of(findUser, findOrder, findStore, requestDto);
+
+        // 생성한 리뷰 저장
+        Reviews savedReview = reviewRepository.save(review);
+
```
요구사항에 배달 완료된 건에 대해서만 리뷰를 달 수 있다는 요구 사항이 있는데, 주문 상태에 대한 검증 로직이 없는 것 같습니다.

주문 상태 Enum 값만 수정해서 공유하고 검증 로직이 있으면 좋을 것 같습니다.

rlqor9688 replied:
반영 완료

rlqor9688 committed:
fix [review] : createReview Transactional 속성 수정

1. Transactional(readOnly = true)에서 readOnly=true 조건 삭제

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/review/service/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,93 @@
+         * </Process>
+         */
+        List<OrderItems> orderItemsList = orderItemsRepository.findByOrdersId(requestDto.getOrdersId());
+        if (!orderItemsList.isEmpty()) {
```
if-else가 없어도 사용할 수 있는 구조 같아요. else에 있는 예외처리를 if에 옮겨 early return 형태로 바꾸는 것이 클린 코드 측면에서 더 적절해 보입니다

nagul2 replied:
오....... 클린코드.. 책읽으셨나여?

anyubinn replied:
책은 아직 안 읽어봤는데 코드리뷰 관련 유튜브 영상 보다가 알게 됐어요!

rlqor9688 replied:
감사합니다! 반영 완료

rlqor9688 committed:
fix [review] : ReviewResponseDto, ReviewServiceImpl 피드백 반영

1. Transactional(readOnly = true)에서 readOnly=true 조건 삭제

2. ReviewReponseDto 메서드명 of -> from 변경

3. ReviewServiceImpl 유저롤 검증/ 유저 인증-Users 메서드로 변경, 유저 찾기 Exception, 배달 완료된 주문 건에만 리뷰를 남길 수 있는 검증 로직 추가, if-else early return 형태로 수정

iop0217 commented:
끝난 코멘트들은 왼쪽 아래 버튼 눌러서 닫으시면 확인하기 편하실 것 같아요

rlqor9688 committed:
Merge branch 'dev' of https://github.com/rlqor9688/igo_delivery into dev

rlqor9688 committed:
Merge branch 'dev' into feature/kiback-review
